theano/scan_module/tests/test_scan.py

```diff
@@ -3895,6 +3916,88 @@ def onestep(x, x_tm4):
         f = theano.function([seq], results[1], updates=updates)
         assert numpy.all(exp_out == f(inp))
 
+    def test_memory_reuse_gpudimshuffle(self):
+        # Test the memory pre-allocation feature in scan when one output is
+        # the result of a GpuDimshuffle (because an optimization in
+        # GpuDimshuffle can cause issues with the memory pre-allocation
+        # where it falsely thinks that a pre-allocated memory region has
+        # been used when it hasn't).
+        from theano.sandbox import cuda
+        if not cuda.cuda_available:
+            raise SkipTest('Optional package cuda disabled')
+
+        def inner_fn(seq1, recurrent_out):
+            temp = seq1 + recurrent_out.sum()
+            output1 = temp.dimshuffle(1, 0)
+            output2 = temp.sum() + recurrent_out
+            return output1, output2
+
+        input1 = theano.tensor.tensor3()
+        init = theano.tensor.tensor3()
+        outputs_info = [None, init]
+
+        out, _ = theano.scan(inner_fn, sequences=[input1],
+                             outputs_info=outputs_info,
+                             mode=mode_with_gpu)
+
+        out1 = out[0].flatten()
+        out2 = out[1].flatten()
+
+        fct = theano.function([input1, init], [out1, out2],
+                              mode=mode_with_gpu)
+
+        floatX = theano.config.floatX
+        output = fct(numpy.ones((2, 1, 1), dtype=floatX),
+                     numpy.ones((1, 1, 1), dtype=floatX))
```

Comment:
True. Truth is, I figured leaving it this way would also check the CPU implementation of Dimshuffle, in case anyone ever decided to apply the same optimization as done in the GPU version.

I should probably update the name and description of the test instead.